object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java

```diff
@@ -543,12 +585,148 @@ private void handleAssignFromRHS(
+   * not. A re-assignment is valid if the called methods type of the lhs before the assignment
+   * satisfies the must-call obligations.
+   *
+   * @param node an assignment to a non-final, owning field
+   * @param newDefs
+   */
+  private void checkReassignmentToField(
+      AssignmentNode node, Set<ImmutableSet<LocalVarWithTree>> newDefs) {
+
+    Node lhsNode = node.getTarget();
+
+    if (!(lhsNode instanceof FieldAccessNode)) {
+      throw new BugInCF(
+          "tried to check reassignment to a field for a non-field node: "
+              + node
+              + " of type: "
+              + node.getClass());
+    }
+
+    FieldAccessNode lhs = (FieldAccessNode) lhsNode;
+
+    Node receiver = lhs.getReceiver();
+    // Check that there is a corresponding resetMustCall annotation, unless this is an
+    // assignment to a field of a newly-declared local variable that can't be in scope
+    // for the containing method.
+    if (!(receiver instanceof LocalVariableNode
+        && isVarInDefs(newDefs, (LocalVariableNode) receiver))) {
+      checkEnclosingMethodIsResetMC(node);
+    }
+
+    MustCallAnnotatedTypeFactory mcTypeFactory =
+        typeFactory.getTypeFactoryOfSubchecker(MustCallChecker.class);
+    AnnotationMirror mcAnno =
+        mcTypeFactory.getAnnotationFromJavaExpression(
+            JavaExpression.fromNode(mcTypeFactory, lhs), node.getTree(), MustCall.class);
+    List<String> mcValues = ValueCheckerUtils.getValueOfAnnotationWithStringArgument(mcAnno);
+
+    if (mcValues.isEmpty()) {
+      return;
+    }
+
+    CFStore cmStoreBefore = typeFactory.getStoreBefore(node);
+    CFValue cmValue = cmStoreBefore == null ? null : cmStoreBefore.getValue(lhs);
+    AnnotationMirror cmAnno =
+        cmValue == null
+            ? typeFactory.top
+            : cmValue.getAnnotations().stream()
+                .filter(anno -> AnnotationUtils.areSameByClass(anno, CalledMethods.class))
+                .findAny()
+                .orElse(typeFactory.top);
+
+    if (!calledMethodsSatisfyMustCall(mcValues, cmAnno)) {
+      Element lhsElement = TreeUtils.elementFromTree(lhs.getTree());
+      checker.reportError(
+          node.getTree(),
+          "required.method.not.called",
+          formatMissingMustCallMethods(mcValues),
+          lhsElement.asType().toString(),
+          " Non-final owning field might be overwritten");
+    }
+  }
+
+  /**
+   * Checks that the method that encloses an assignment is marked with @ResetMustCall annotation
+   * whose target is the object whose field is being re-assigned.
+   *
+   * @param node an assignment node whose lhs is a non-final, owning field
+   */
+  private void checkEnclosingMethodIsResetMC(AssignmentNode node) {
+    Node lhs = node.getTarget();
+    if (!(lhs instanceof FieldAccessNode)) {
+      return;
+    }
+
+    String receiverString = receiverAsString((FieldAccessNode) lhs);
+
+    TreePath currentPath = typeFactory.getPath(node.getTree());
```

Comment:
Is there any way to get the top-level tree for a CFG given some node in that CFG, which should be the desired `MethodTree`?  Just curious; I'm getting that getting the path for a tree might be expensive (requiring a search of the AST for the whole compilation unit)

Reply:
There might be, but the path is needed for viewpoint adaptation later in the method regardless so computing it is unavoidable.

It might be possible to defer computing the path until after we check for the presence of an `@ResetMustCall` annotation on the declaration, if we were able to get the `MethodTree` another way - which would probably make it worthwhile.

Reply:
Ok.  We don't need to block landing on this, but maybe we should add a comment if we think there may be a better way to do it in the future.

Reply:
I'll add a comment.